examples/parser/Parser/CSV.roc

```diff
@@ -0,0 +1,196 @@
+        Err (ParsingIncomplete rest) ->
+            restStr = Parser.Str.strFromRaw rest
+
+            Err (SyntaxError restStr)
+
+        Err (ParsingFailure str) ->
+            Err (ParsingFailure str)
+
+        Ok csvData ->
+            when parseCSV csvParser csvData is
+                Err (ParsingFailure str) ->
+                    Err (ParsingFailure str)
+
+                Err (ParsingIncomplete problem) ->
+                    Err (ParsingIncomplete problem)
+
+                Ok vals ->
+                    Ok vals
+
+## Attempts to parse an `a` from a `CSV` datastructure (a list of lists of bytestring-fields).
+parseCSV : Parser CSVRecord a, CSV -> Result (List a) [ParsingFailure Str, ParsingIncomplete CSVRecord]
+parseCSV = \csvParser, csvData ->
+    csvData
+    |> List.mapWithIndex (\recordFieldsList, index -> { record: recordFieldsList, index: index })
+    |> List.walkUntil (Ok []) \state, { record: recordFieldsList, index: index } ->
+        when parseCSVRecord csvParser recordFieldsList is
+            Err (ParsingFailure problem) ->
+                indexStr = Num.toStr (index + 1)
+                recordStr = recordFieldsList |> List.map strFromRaw |> List.map (\val -> "\"\(val)\"") |> Str.joinWith ", "
+                problemStr = "\(problem)\nWhile parsing record no. \(indexStr): `\(recordStr)`"
+
+                Break (Err (ParsingFailure problemStr))
+
+            Err (ParsingIncomplete problem) ->
+                Break (Err (ParsingIncomplete problem))
+
+            Ok val ->
+                state
+                |> Result.map (\vals -> List.append vals val)
+                |> Continue
+
+## Attempts to parse an `a` from a `CSVRecord` datastructure (a list of bytestring-fields)
+##
+## This parser succeeds when all fields of the CSVRecord are consumed by the parser.
+parseCSVRecord : Parser CSVRecord a, CSVRecord -> Result a [ParsingFailure Str, ParsingIncomplete CSVRecord]
+parseCSVRecord = \csvParser, recordFieldsList ->
+    parse csvParser recordFieldsList (\leftover -> leftover == [])
+
+## Wrapper function to combine a set of fields into your desired `a`
+##
+## ## Usage example
+##
+## >>> record (\firstName -> \lastName -> \age -> User {firstName, lastName, age})
+## >>> |> field string
+## >>> |> field string
+## >>> |> field nat
+##
+record : a -> Parser CSVRecord a
+record = Parser.Core.const
+
+## Turns a parser for a `List U8` into a parser that parses part of a `CSVRecord`.
+field : Parser RawStr a -> Parser CSVRecord a
+field = \fieldParser ->
+    buildPrimitiveParser \fieldsList ->
+        when List.get fieldsList 0 is
+            Err OutOfBounds ->
+                Err (ParsingFailure "expected another CSV field but there are no more fields in this record")
+
+            Ok rawStr ->
+                when Parser.Str.parseRawStr fieldParser rawStr is
+                    Ok val ->
+                        Ok { val: val, input: List.dropFirst fieldsList }
```

Comment:
Not terribly important right now, but I wonder if it would be better to
- keep an index into the whole list here, or
- reverse the list and operate on it in reverse order
rather than continuously dropping the first element, which takes linear time (quadratic over all field parsing) because Roc `List`s are not linked lists

Reply:
Reason I had not done this before was to get something simple (that was closer to the reference material) working fast. This is indeed a very nice optimization possibility.